syntax = "proto3";
package synthpop;

message Population {
    repeated Household households = 1;
    repeated Person people = 2;

    // This is filled out per Activity enum value. The exception is for
    // `Activity::Home`; see `households` for that.
    map<int32, VenueList> venues_per_activity = 3;

    map<string, InfoPerMSOA> info_per_msoa = 4;
}

message Household {
    uint64 id = 1;
    string msoa = 2;
    /// An ID from the original data, kept around for debugging
    int64 orig_hid = 3;
    // These IDs index into `Population::people`
    // TODO These're redundant; worth putting here?
    repeated uint64 members = 4;
}

message VenueList {
    repeated Venue venues = 1;
}

message InfoPerMSOA {
    repeated Point shape = 1;
    uint64 population = 2;
    // All building centroids within this MSOA. May be empty.
    repeated Point buildings = 3;
}

// In WGS84
message Point {
    float longitude = 1;
    float latitude = 2;
}

message Person {
    uint64 id = 1;
    uint64 household = 2;
    // This is the centroid of the household's MSOA. It's redundant to store it per person, but
    // very convenient.
    Point location = 3;
    // An ID from the original data, kept around for debugging
    int64 orig_pid = 4;
    // The Standard Industry Classification for where this person works. 0 means none.
    uint64 sic1d07 = 5;

    uint32 age_years = 6;

    Health health = 7;

    repeated Flows flows_per_activity = 8;
}

message Health {
    Obesity obesity = 1;
    bool has_cardiovascular_disease = 2;
    bool has_diabetes = 3;
    bool has_high_blood_pressure = 4;
}

// Per activity, a list of venues where this person is likely to go do that activity.
message Flows {
    Activity activity = 1;
    repeated Flow flows = 2;
    // These sum to 1 for all `flows_per_activity`, representing a fraction of a day
    double activity_duration = 3;
}

message Flow {
    uint64 venue_id = 1;
    // The probabilities sum to 1 for all `flows`
    double weight = 2;
}

// Represents a place where people do an activity
message Venue {
    // IDs are reused across different activities; both this id and an Activity
    // is needed to lookup from venues_per_activity.
    uint64 id = 1;
    Activity activity = 2;

    Point location = 3;
    // This only exists for PrimarySchool and SecondarySchool. It's a
    // https://en.wikipedia.org/wiki/Unique_Reference_Number. 0 means none.
    uint64 urn = 4;
}

enum Activity {
    RETAIL = 0;
    PRIMARY_SCHOOL = 1;
    SECONDARY_SCHOOL = 2;
    HOME = 3;
    WORK = 4;
    NIGHTCLUB = 5;
}

enum Obesity {
    NORMAL = 0;
    OVERWEIGHT = 1;
    OBESE_1 = 2;
    OBESE_2 = 3;
    OBESE_3 = 4;
}
